client/my-sites/stats/activity-log-search/activity-log-search-groups.jsx

```diff
@@ -0,0 +1,109 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import classNames from 'classnames';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { pull, union } from 'lodash';
+import React, { Component } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import getActivityLogFilter from 'state/selectors/get-activity-log-filter';
+import Gridicon from 'gridicons';
+import { updateFilter } from 'state/activity-log/actions';
+
+// Note: this data-set will eventually come from the API
+const groups = [
+	{ slug: 'post', title: 'Posts and Pages', icon: 'posts' },
+	{ slug: 'user', title: 'Users', icon: 'user' },
+	{ slug: 'comment', title: 'Comments', icon: 'comment' },
+	{ slug: 'attachment', title: 'Media and Attachments', icon: 'attachment' },
+	{ slug: 'setting', title: 'Settings', icon: 'cog' },
+	{ slug: 'term', title: 'Categories and Tags', icon: 'tag' },
+];
+
+class ActivityLogSearchGroups extends Component {
+	onGroupClick = event => {
+		const { filter, siteId } = this.props;
+		const currentState = event.target.getAttribute( 'class' );
+		const group = event.target.getAttribute( 'data-group' );
```

Comment:
also known as `const { group } = event.target.dataset;`

might be worth passing the include/exclude prop here too. it's not _bad_ or _broken_ passing it via the CSS selectors, but I think those will turn out to be best left as visual indicators that people feel more free to play around with than specific data values.

```js
const { groupSlug, intent } = event.target.dataset;

…

<button
	data-group-slug={ group.slug }
	data-intent={ someCondition ? 'include' : 'exclude' }
>
```

interestingly enough I detected a race condition here and I wonder if we can get around it. there is the possibility that the DOM changes between the time of our click and the time this handler gets called; there's also the possibility that `this.props` changes. I'm not sure we can avoid it unless we go crazy, but _I think_ we an mitigate some of it by _not passing_ the values which end up being inferred from props anyway - or maybe, we're shrinking the size of the race condition here. (note that in the code in this PR it's possible to be both included and excluded at the same time)

```js
onGroupClick = ( { target: { dataset: { groupSlug } } } ) => {
	const { filter, siteId } = this.props;
	const { group: filterIncludes, notGroup: filterExcludes } = filter;

	const wasIncluded = filterIncludes.includes( groupSlug );
	const wasExcluded = filterExcludes.includes( groupSlug );

	if ( wasIncluded && wasExcluded ) {
		// bail - fix "higher up the stack"
		return;
	}

	const includesUpdates = wasIncluded && {
		group: without( filter.group, groupSlug ),
		notGroup: union( filter.notGroup, [ groupSlug ] ),
	};

	const excludesUpdates = wasExcluded && {
		group: union( filter.group, [ groupSlug ] ),
		notGroup: without( filter.group, groupSlug ),
	};

	const defaultUpdates = ! ( wasIncluded || wasExcluded ) && {
		group: union( filter.group, [ groupSlug ] ),
	};

	this.props.updateFilter(
		siteId,
		{
			...filter,
			...( includesUpdates || {} ),
			...( excludesUpdates || {} ),
			...( defaultUpdates || {} ),
		}
	);
}
```

Since this comment grew I'll add notes here instead of below.

 - there's a symmetry I added back in that felt missing in the given code whereby the `included` branch adjusted both the `group` _and_ the `notGroup` properties while the other two branches only adjusted one each
 - `lodash.pull()` mutates its array and will introduce render bugs into React - `lodash.without()` is a non-mutating alternative
 - `lodash.union()` will return a list when given an `undefined` input, `lodash.without()` will return an empty list when no more elements remain. it's safe to use these without the extra conditionals. we only need to make sure that our logic understands that `[]` is falsey

seeing this code makes me think that there's more work to do yet on the data structure we're using. why can we have filters both on the includes and excludes list at the same time? why do we have to move an item out of one list and then into another? why isn't it one operation?

these are tri-state buttons aren't they? I know it's convenient to eventually send these as two separate `group` and `not-group` properties through the URL but would it make sense to store them internally as tri-state values? could even use the [tri-state checkbox](https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-2/checkbox-2.html) in order to make them more accessible: 'include', 'exclude', 'neither'

```js
// psuedo-code
this.props.updateFilter(
	siteId,
	{
		...filter,
		groups: {
			...filter.groups,
			[ groupSlug ]: ( 'include' | 'exclude' )
		}
	}
);
```

or even in a flat world
```js
const { filterGroups } = this.props;

if ( ! ( shouldInclude || shouldExclude ) ) {
	const { [ groupSlug ]: removed, ...remained } = filterGroups;

	return remained;
}

return {
	...filterGroups,
	[ groupSlug ]: shouldInclude ? 'include' : 'exclude',
};
```

questions…